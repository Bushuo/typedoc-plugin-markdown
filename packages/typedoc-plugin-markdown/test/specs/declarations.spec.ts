import * as Handlebars from 'handlebars';

import { TestApp } from '../test-app';

describe(`Declarations:`, () => {
  let testApp: TestApp;

  let template: Handlebars.TemplateDelegate;

  beforeAll(async () => {
    testApp = new TestApp(['declarations.ts']);
    await testApp.bootstrap();
    TestApp.stubPartials(['member.sources']);
    TestApp.stubHelpers([]);

    template = TestApp.getPartial('member.declaration');
  });

  test(`should compile a variable with default value`, () => {
    expect(
      TestApp.compileTemplate(
        template,
        testApp.findReflection('stringWithDefaultValueDeclaration'),
      ),
    ).toMatchSnapshot();
  });

  test(`should compile a an udefined`, () => {
    expect(
      TestApp.compileTemplate(
        template,
        testApp.findReflection('undefinedNumberDeclaration'),
      ),
    ).toMatchSnapshot();
  });

  test(`should compile object literal declaration`, () => {
    expect(
      TestApp.compileTemplate(
        template,
        testApp.findReflection('objectLiteralDeclaration'),
      ),
    ).toMatchSnapshot();
  });

  test(`should compile object literal cast as a const`, () => {
    expect(
      TestApp.compileTemplate(
        template,
        testApp.findReflection('objectLiteralAsConstDeclaration'),
      ),
    ).toMatchSnapshot();
  });

  test(`should compile type literal declaration`, () => {
    expect(
      TestApp.compileTemplate(
        template,
        testApp.findReflection('typeLiteralDeclaration'),
      ),
    ).toMatchSnapshot();
  });

  test(`should compile type literal declaration`, () => {
    expect(
      TestApp.compileTemplate(
        template,
        testApp.findReflection('typeLiteralDeclarationWithFunction'),
      ),
    ).toMatchSnapshot();
  });

  test(`should compile declaration with double underscores in name and value`, () => {
    expect(
      TestApp.compileTemplate(
        template,
        testApp.findReflection('__DOUBLE_UNDERSCORES_DECLARATION__'),
      ),
    ).toMatchSnapshot();
  });

  test(`should compile any function type`, () => {
    expect(
      TestApp.compileTemplate(
        template,
        testApp.findReflection('AnyFunctionType'),
      ),
    ).toMatchSnapshot();
  });

  test(`should not escape charcters within markdown code`, () => {
    expect(
      TestApp.compileTemplate(
        template,
        testApp.findReflection('SpecialCharacters'),
      ),
    ).toMatchSnapshot();
  });

  test(`should compile enum delcaration`, () => {
    expect(
      TestApp.compileTemplate(
        template,
        testApp.findReflection('EnumDeclarations').children[0],
      ),
    ).toMatchSnapshot();
  });
});
