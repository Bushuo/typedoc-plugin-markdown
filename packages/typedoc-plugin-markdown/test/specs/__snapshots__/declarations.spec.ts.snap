// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Declarations: should compile a an udefined 1`] = `
"• \`Let\` **undefinedNumberDeclaration**: *number*

[partial: member.sources]
"
`;

exports[`Declarations: should compile a variable with default value 1`] = `
"• \`Const\` **stringWithDefaultValueDeclaration**: *variable*= 'variable'

[partial: member.sources]
"
`;

exports[`Declarations: should compile any function type 1`] = `
"Ƭ **AnyFunctionType**<A\\\\>: (...\`input\`: *any*[]) => A

#### Type parameters:

Name | Default |
------ | ------ |
\`A\` | *any* |

[partial: member.sources]
"
`;

exports[`Declarations: should compile declaration with double underscores in name and value 1`] = `
"• \`Const\` **\\\\_\\\\_DOUBLE\\\\_UNDERSCORES\\\\_DECLARATION\\\\_\\\\_**: *typeof* [*\\\\_\\\\_DOUBLE\\\\_UNDERSCORES\\\\_DECLARATION\\\\_\\\\_*](../modules.md#__double_underscores_declaration__)

[partial: member.sources]
"
`;

exports[`Declarations: should compile enum delcaration 1`] = `
"• **Down**: = \\"DOWN\\"

[partial: member.sources]
"
`;

exports[`Declarations: should compile object literal cast as a const 1`] = `
"• \`Const\` **objectLiteralAsConstDeclaration**: object

#### Type declaration:

Name | Type | Value |
------ | ------ | ------ |
\`Prop1\` | *Prop1* | 'Prop1' |
\`Prop2\` | *Prop1* | 'Prop1' |
\`Prop3\` | *Prop1* | 'Prop1' |

[partial: member.sources]
"
`;

exports[`Declarations: should compile object literal declaration 1`] = `
"• \`Const\` **objectLiteralDeclaration**: object

#### Type declaration:

Name | Type | Value |
------ | ------ | ------ |
\`valueA\` | *number* | 100 |
\`valueB\` | *boolean* | true |
\`valueC\` | object | {} |
\`valueX\` | object | { \`valueA\`: *number*[] ; \`valueZ\`: *string* = 'foo' } |
\`valueY\` | function | () => *string* |
\`valueZ\` | *string* | 'foo' |

[partial: member.sources]
"
`;

exports[`Declarations: should compile type literal declaration 1`] = `
"• \`Let\` **typeLiteralDeclaration**: object

#### Type declaration:

Name | Type | Value | Description |
------ | ------ | ------ | ------ |
\`valueA?\` | *number* | *number* | Comments for valueA   |
\`valueB?\` | *boolean* | *boolean* | - |
\`valueX\` | object | { \`valueA\`: *number*[] ; \`valueY\`: (\`z\`: *string*) => { \`a\`: *string* ; \`b\`: *string*  } ; \`valueZ\`: *string*  } | Comment for valueX   |
\`valueY\` | function | () => *string* | - |
\`valueZ\` | *string* | *string* | Comment for valueZ   |

[partial: member.sources]
"
`;

exports[`Declarations: should compile type literal declaration 2`] = `
"• \`Let\` **typeLiteralDeclarationWithFunction**: () => *string*

#### Type declaration:

Name | Type | Value |
------ | ------ | ------ |
\`valueZ\` | *string* | *string* |

[partial: member.sources]
"
`;

exports[`Declarations: should not escape charcters within markdown code 1`] = `
"Ƭ **SpecialCharacters**: { \`this_prop_has_underscores\`: *number* ; \`this|prop|has|the|pipe|character\`: *string*  }

#### Type declaration:

Name | Type |
------ | ------ |
\`this_prop_has_underscores\` | *number* |
this\\\\|prop\\\\|has\\\\|the\\\\|pipe\\\\|character | *string* |

[partial: member.sources]
"
`;
